src/main/java/io/pinecone/clients/Pinecone.java

```diff
@@ -149,14 +152,27 @@ public void deleteCollection(String collectionName) throws PineconeException {
         }
     }
 
-    public Index createIndexConnection(String indexName) {
-        PineconeConnection connection = new PineconeConnection(config, indexName);
-        return new Index(connection);
+    public Index getIndexConnection(String indexName) {
+        config.setHost(getIndexHost(indexName));
+        PineconeConnection connection = getConnection(indexName);
+        return new Index(this, connection, indexName);
+    }
+
+    public AsyncIndex getAsyncIndexConnection(String indexName) {
+        PineconeConnection connection = getConnection(indexName);
+        return new AsyncIndex(this, connection, indexName);
+    }
+
+    private PineconeConnection getConnection(String indexName) {
+        return connectionsMap.computeIfAbsent(indexName, key -> new PineconeConnection(config));
```

Comment:
@ssmith-pc made a great suggestion to use computeIfAbsent() which will compute/create a new connection if indexName is not found in the map. Initially, it was an if-else loop so we would have to either acquire and release the locks to make the process of checking if the key doesn't exists and creating the connection object atomic.

Reply:
Where is the `config` object coming from?